### PR TITLE
Use improvements

### LINKED ISSUE #3031: Improve error message for `use` functions with wrong last argument
Consider the following snippet of code: 
```gleam
pub fn main() {
  use a <- io.println
  1
}
```
The error produced by the compiler looks like this:
```txt
error: Type mismatch
  ┌─ /src/main.gleam:5:3
  │
5 │   1
  │   ^

Expected type:

    String

Found type:

    fn(a) -> Int
```
I think it would be way more useful if the error was something like this:
```txt
error: Type mismatch
  ┌─ /src/main.gleam
  │
4 │   use a <- io.println
  │            ^^^^^^^^^^

This function cannot be used on the right hand side of a `use`.
I was expecting its last argument to be a callback function with type:

  fn(a) -> String

But its actual type is:

  String
  
See: https://tour.gleam.run/advanced-features/use/
```
However, if the last argument is actually a callback function the error message could be worded differently, just highlighting the type mismatch:
```text
error: Type mismatch
  ┌─ /src/main.gleam
  │
4 │   use a, b <- result.try(result)
  │               ^^^^^^^^^^

This function accepts a callback function with a different type. 
Expected type:

  fn(Int) -> Result(a, String)

Found type:

  fn(a, b) -> Int
```

### PR DESCRIPTION
**Pull request by giacomocavalieri**
This PR improves on various use error messages.
Closes #3031, #3032 and #3033

### CODE REVIEW

File: compiler-core/src/type_/expression.rs [hunk 1]
```diff
@@ -2344,20 +2398,42 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
...
                     location,
                     implicit,
                 } = arg;
-                let value = self.infer_call_argument(value, typ.clone())?;
+
+                // If we're typing a `use` call then the last argument is the
+                // use callback and we want to treat it differently to report
+                // better errors.
+                let is_last_argument = i == args_count - 1;
+                let argument_kind = match (kind, is_last_argument) {
```
Comment (lpil) [inline compiler-core/src/type_/expression.rs:2421]: Can we match on just `kind` and use `is_last_argument` in a guard please

File: compiler-core/src/type_/expression.rs [hunk 2]
```diff
@@ -528,13 +568,22 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
...
+            call.arguments,
+            call_location,
+            CallKind::Use {
+                assignments_location,
+                last_statement_location,
+            },
+        )?;
 
-        Ok(Statement::Expression(call))
+        Ok(Use {
```
Comment (lpil) [inline compiler-core/src/type_/expression.rs:581]: This has radically changed the meaning and data of `Use`- now the callback code is within the call in the `Use`. Can we document this somewhere? Or is there some way to make it clearer?

File: compiler-core/src/type_/expression.rs [hunk 3]
```diff
@@ -2410,7 +2487,19 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
...
+                assignments_location,
+                last_statement_location,
+            } => e.into_use_unify_error(
+                function_location,
+                assignments_location,
+                last_statement_location,
+                value.location(),
+            ),
+            ArgumentKind::Regular => convert_unify_error(e, value.location()),
+        })?;
```
Comment (lpil) [inline compiler-core/src/type_/expression.rs:2502]: Could you extract this into a method please? I'm finding it a bit hard to see the logic as the majority of the code is type conversion boilerplate.

File: compiler-core/src/type_/expression.rs [hunk 4]
```diff
@@ -2705,10 +2794,10 @@ struct UseAssignments {
 }
 
 impl UseAssignments {
-    fn from_use_expression(sugar_assignments: Vec<UseAssignment>) -> UseAssignments {
+    fn from_use_expression(sugar_assignments: &[UseAssignment]) -> UseAssignments {
```
Comment (lpil) [inline compiler-core/src/type_/expression.rs:2797]: There's a bunch of cloning here which we didn't have to do before. Why is that?

**Review by lpil — COMMENTED** [review]
Wonderful work! I am very impressed!

I've left some comments in-line, mostly minor code readability things. I've also left one question about some cloning which I couldn't quite figure out.

File: compiler-core/src/error.rs [hunk 5]
```diff
@@ -2725,7 +2725,166 @@ Rename or remove one of them.",
...
+                            path: path.clone(),
+                            src: src.clone(),
+                            extra_labels: vec![],
+                        }),
+                    }
+                },
+
+                TypeError::UseFnDoesntTakeCallback { location, actual_type: Some(actual) } => {
+                    let mut printer = Printer::new();
+                    let text = wrap_format!("The function on the right hand side of `<-` has to take a callback function as its last argument. But the last argument of this function has type:
```
[the remark this answers is not in the document]
Comment (giacomocavalieri) [reply compiler-core/src/error.rs:2827]: Aha! I was wondering why it stopped working all of a sudden 😆

File: compiler-core/src/type_/environment.rs [hunk 6]
```diff
@@ -726,19 +726,49 @@ pub fn unify(t1: Arc<Type>, t2: Arc<Type>) -> Result<(), UnifyError> {
...
+                    .map_err(|_| UnifyError::CouldNotUnify {
+                        expected: t1.clone(),
+                        given: t2.clone(),
+                        situation: Some(UnifyErrorSituation::FunctionsMismatch {
+                            reason: FunctionsMismatchReason::Results {
+                                one: retrn1.clone(),
+                                other: retrn2.clone(),
+                            },
+                        }),
+                    })
```
[the remark this answers is not in the document]
Comment (giacomocavalieri) [reply compiler-core/src/type_/environment.rs:759]: Oops!

File: compiler-core/src/type_/expression.rs [hunk 7]
```diff
@@ -505,6 +537,14 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
         }
 
         let statements = Vec1::try_from_vec(statements).expect("safe: todo added above");
+        // We need this to report good error messages in case there's a type error
+        // in use. We consider `use` to be the last statement of a block since
+        // it consumes everything that comes below it and returns a single value.
+        let last_statement_location = statements
+            .iter()
+            .find_or_last(|e| e.is_use())
```
[the remark this answers is not in the document]
Comment (giacomocavalieri) [reply compiler-core/src/type_/expression.rs:545]: The cool part is I discovered completely by accident 😆

File: compiler-core/src/error.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (lpil) [reply compiler-core/src/error.rs:2827]: It took me forever to work that one out. It's really hard to find the lines that cause the problem when it happens too

**Comment by lpil** [comment]
I think you're working on this so I've made it a draft so I know not to look yet in my notifications list

File: compiler-core/src/type_/expression.rs (hunk 2 above)
Comment (giacomocavalieri) [reply compiler-core/src/type_/expression.rs:581]: Ok, I'm adding some comments!

**Review by lpil — APPROVED** [review]
Incredible work as per usual! This is going to make such a huge difference to how easy it is to debug `use`

**Pull request merged**